### PR TITLE
CORE-19374: Prevent failed states from being processed in the event mediator

### PR DESCRIPTION
**Pull request by JamesHR3**
### Problem statement

It is possible during event processing for a failure to occur. This will usually happen due to a processing timeout, either in handling a synchronous request or in completing the full event processing. These cases result in the state being marked as processing failed. This occurs even if the state had never been created, as the client may wish to inspect which states failed processing.

In some scenarios this can lead to the state value not containing the expected object type. This leads to a deserialization problem in the event processing code if a new set of events is polled for on the same key.

### Solution

Prevent the processing of events for states that are marked as failed.

To do this, the `ConsumerProcessor` now does some initial filtering to remove any states marked as failed from being eligible for processing, alongside the corresponding events for that state. To that end, a new `EventProcessingInput` class has been defined to make it clearer what actually will get processed. These are generated from the input events and states and provided to the event processor.

### CODE REVIEW

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/processor/ConsumerProcessor.kt [hunk 1]
```diff
@@ -106,27 +107,28 @@ class ConsumerProcessor<K : Any, S : Any, E : Any>(
...
-                }.map { (future, keys) ->
+                    Pair(future, group)
+                }.map { (future, group) ->
                     try {
                         future.getOrThrow(Duration.ofMillis(EVENT_PROCESSING_TIMEOUT_MILLIS))
                     } catch (e: TimeoutException) {
-                        keys.associateWith {
-                            val oldState = statesToProcess[it.toString()]
-                            val state = stateManagerHelper.failStateProcessing(it.toString(), oldState)
+                        group.mapValues { (key, input) ->
```
Comment (LWogan) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/processor/ConsumerProcessor.kt:129]: I think we should add a warn log here. I'm not sure we log timeouts anywhere?

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt [hunk 2]
```diff
@@ -47,13 +47,19 @@ class StateManagerHelper<S : Any>(
...
+     * allows clients to detect issues with any failed processing.
+     */
     fun failStateProcessing(key: String, originalState: State?) : State {
         val newMetadata = (originalState?.metadata?.toMutableMap() ?: mutableMapOf()).also {
             it[PROCESSING_FAILURE] = true
         }
         return State(
             key,
-            byteArrayOf(),
+            originalState?.value ?: byteArrayOf(),
```
Comment (LWogan) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt:62]: I think byteArrayOf() will result in some deserialization warn logs in TimeoutEventCleanupProcessor when we try to deserialize the state.

**Review by LWogan — COMMENTED** [review]
Couple of initial comments while they're fresh in my head.

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/processor/ConsumerProcessor.kt (hunk 1 above)
Comment (JamesHR3) [reply libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/processor/ConsumerProcessor.kt:129]: We have a separate ticket to do this logging under. This isn't the only place we're missing a few things, so I'd like to do it all under the same umbrella.

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt (hunk 2 above)
Comment (JamesHR3) [reply libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt:62]: `FlowTimeoutTaskProcessor` currently prevents this by doing a search for a new `STATE_TYPE` key, which is set to `Checkpoint` for flows. This is a workaround for exactly this problem. What we actually want is to be able to clean up things where this value is not set, but some extra work is required to get that set up correctly. We're using that as a stop gap until a proper solution is implemented.

**Review by jujoramos — APPROVED** [review]

**Review by LWogan — APPROVED** [review]

**Pull request merged**